File: go/example_host.go
package main

import (
	"os"
	"os/signal"

	"github.com/chadbohannan/application-layer-network/go/aln"
)

func main() {
	alnHostAddress := aln.AddressType(1)
	pingServiceID := uint16(1)
	router := aln.NewRouter(alnHostAddress)
	tcpHost := aln.NewTcpChannelHost("localhost", 8000)
	go tcpHost.Listen(func(newChannel aln.Channel) {
		router.AddChannel(newChannel)
	})

	// create the ping service to send packets back where they came from
	router.RegisterService(pingServiceID, func(packet *aln.Packet) {
		router.Send(&aln.Packet{
			DestAddr:  packet.SrcAddr,
			ContextID: packet.ContextID,
			Data:      []byte("pong!"),
		})
	})

	// wait for ^C
	c := make(chan os.Signal, 1)
	signal.Notify(c, os.Interrupt)
	<-c
}

File: python/example_host.py
import selectors, signal, socket, sys, time
from aln.parser import Packet
from aln.tcphost import TcpHost
from aln.router import Router
from threading import Lock


def main():
    sel = selectors.DefaultSelector() # application event loop
    router = Router(sel, "python-host-1") # TODO dynamic address allocation protocol
    router.start()

    def ping_handler(packet):
        print("ping from [{0}]".format(packet.srcAddr))
        router.send(Packet(
            destAddr=packet.srcAddr,
            contextID=packet.contextID,
            data="pong"
        ))

    router.register_service("ping", ping_handler)

    def on_connect(tcpChannel, addr):
        print('accepted connection from', addr)
        router.add_channel(tcpChannel)

    tcpHost = TcpHost()
    tcpHost.listen(sel, 'localhost', 8181, on_connect)
    
    def signal_handler(signal, frame): # listen for ^C
        router.close()
        sel.close()
        lock.release()
    signal.signal(signal.SIGINT, signal_handler)

    lock = Lock()
    lock.acquire()
    lock.acquire()
    lock.release()

if __name__ == "__main__":
    main()
